### PR TITLE
[luci/import] Add VectorWrapper helper struct

### PR DESCRIPTION
**Pull request by m-bronnikov**
This commit adds VectorWrapper struct which helps to use `flatbuffers::Vector<>` from direct-access API.

ONE-DCO-1.0-Signed-off-by: Maksim Bronnikov max120199@gmail.com

---------------------

For: #7886
Draft: #7901

### CODE REVIEW

**Comment by seanshpark** [comment]
@m-bronnikov , as this class is introduce, we may need some unit tests for this.
I couldn't see any unit tests modified in thje draft PR. Sorry to point this out now.
As our SE requires there should be same or more negative tests, please add same or more negative unit tests.

**Comment by m-bronnikov** [comment]
> @m-bronnikov , as this class is introduce, we may need some unit tests for this. I couldn't see any unit tests modified in thje draft PR. Sorry to point this out now. As our SE requires there should be same or more negative tests, please add same or more negative unit tests.

Can I create *CircleReader.test.cpp* and add some tests to this file for this purpose?

**Comment by seanshpark** [comment]
> Can I create CircleReader.test.cpp and add some tests to this file for this purpose?

Yes, please :)

**Comment by seanshpark** [comment]
We may later introduce `VectorWrapper` in another file.

**Comment by m-bronnikov** [comment]
> > @m-bronnikov , as this class is introduce, we may need some unit tests for this. I couldn't see any unit tests modified in thje draft PR. Sorry to point this out now. As our SE requires there should be same or more negative tests, please add same or more negative unit tests.
> 
> Can I create _CircleReader.test.cpp_ and add some tests to this file for this purpose?

@seanshpark Done =)

File: compiler/luci/import/src/CircleReader.test.cpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Import/CircleReader.h"
+
+#include <gtest/gtest.h>
+
+TEST(VectorWrapperTest, basic_pattern_P)
```
Comment (seanshpark) [inline compiler/luci/import/src/CircleReader.test.cpp:21]: `_P` is not needed for normal(positive)

File: compiler/luci/import/src/CircleReader.test.cpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+TEST(VectorWrapperTest, null_pointer_P)
+{
+  flatbuffers::Vector<int32_t> *vec_pointer = nullptr;
+  auto const wrapper = luci::wrap(vec_pointer);
+
+  ASSERT_TRUE(wrapper.null());
+  ASSERT_TRUE(wrapper.empty());
+}
+
+TEST(VectorWrapperTest, prohibited_access_N)
```
Comment (seanshpark) [inline compiler/luci/import/src/CircleReader.test.cpp:45]: Please use `_NEG` for negative tests

File: compiler/luci/import/src/CircleReader.test.cpp (hunk 1 above)
Comment (m-bronnikov) [reply compiler/luci/import/src/CircleReader.test.cpp]: Removed.

File: compiler/luci/import/src/CircleReader.test.cpp (hunk 2 above)
Comment (m-bronnikov) [reply compiler/luci/import/src/CircleReader.test.cpp]: Done.

File: compiler/luci/import/src/CircleReader.test.cpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  ASSERT_TRUE(wrapper.empty());
+}
+
+TEST(VectorWrapperTest, prohibited_access_NEG)
+{
+  flatbuffers::Vector<uint8_t> *vec_pointer = nullptr;
+  auto const wrapper = luci::wrap(vec_pointer);
+
+  ASSERT_ANY_THROW(wrapper.at(0));
+}
```
Comment (seanshpark) [inline compiler/luci/import/src/CircleReader.test.cpp:50]: There are two positive and one negative. please add more negative tests.
`count(POS) <= count(NEG)`

**Comment by seanshpark** [comment]
I wrote
> As our SE requires there should be same or more negative tests, please add same or more negative unit tests.

**Comment by seanshpark** [comment]
Our repo is in boundary in positive test count ~= negative test count.
if positive test count > negative test count (counted for new test cases from some time ago),
SE will not grant our release.

**Comment by m-bronnikov** [comment]
> Our repo is in boundary in positive test count ~= negative test count. if positive test count > negative test count (counted for new test cases from some time ago), SE will not grant our release.

Ahh, now I understand, thank you for clarifying =)

**Comment by m-bronnikov** [comment]
I added another `_NEG` test.

File: compiler/luci/import/src/CircleReader.cpp [hunk 4]
```diff
@@ -350,4 +350,61 @@ bool CircleReader::select_subgraph(uint32_t sgindex)
...
+}
+
+template <typename T> typename VectorWrapper<T>::iterator VectorWrapper<T>::end() const
+{
+  return null() ? begin() : _vector->end();
+}
+
+template <typename T> typename VectorWrapper<T>::value_type VectorWrapper<T>::at(uint32_t i) const
+{
+  if (i >= size())
```
Comment (binarman) [inline compiler/luci/import/src/CircleReader.cpp:381]: Add check for `null()`, otherwise could segfault on line 387.
This is also an opportunity for another negative test =)

**Review by seanshpark — APPROVED** [review]
LGTM thank you!
